kernel/copy.masm

```diff
@@ -0,0 +1,265 @@
+
+  # check if the odd bit is set
+  push.1 u32checked_and
+end
+
+#! move_from_tape_to_memory_even_words(
+#!   A: Word,
+#!   B: Word,
+#!   C: Word,
+#!   write_ptr: address,
+#!   words: felt,
+#! ) -> [A', B', C', write_ptr']
+#!
+#! This is an optimized version of the inner loop.
+#!
+#! Given an initialized RPO state represented by [A, B, C], moves an even amount
+#! `words` from the advice tape sequentially memory starting at address
+#! `write_ptr`. 
+#!
+#! Note:
+#! - words must be even and greater than zero
+#! - The words A,B,C compose the state of the RPO, these words must have been
+#! properly initialized, including setting Capacity to [0,0,0,1] if the complete
+#! copy is for an odd number of words
+#
+#! input: [A, B, C, write_ptr, words, ...]
+#! output: [A, B, C, write_ptr, ...]
+#! cycles: 7 + 8*words
+proc.move_from_tape_to_memory_even_words.0
+  # LOOP: [A, B, C, write_ptr, words, ...]
+  # while(words>0)
+  dup.13 neq.0 while.true
+    # stack: [A', B', C', write_ptr', words, ...]
+    adv_pipe
+
+    # move `words` counter up
+    # stack: [[write_ptr', words, x, x], A', B', C', ...]
+    swapw.3
+
+    # stack: [[words, write_ptr', x, x], A', B', C', ...]
+    movup.1
+
+    # optimization: `words` is required to be even and greater-than zero for
+    # this procedure to be used. Here it is possible to do a division instead
+    # of a range check, so the test is a single cycle
+    # stack: [[words/2, write_ptr', x, x], A', B', C', ...]
+    div.2
+
+    # move `words` counter down
+    # LOOP: [A', B', C', write_ptr', words/2, ...]
+    movdn.1 swapw.3
+
+    # while(words>0)
+    dup.13 neq.0
+  end
+
+  # stack: [A', B', C', write_ptr', ...]
+  movup.13 drop
+end
+
+#! move_from_tape_to_memory(words: felt, write_ptr: address) -> [rpo_hash, write_ptr']
+#!
+#! Moves the `words` words from the advice tape to memory, starting at
+#! `write_ptr` address, the hash of the copied data is left on the stack for
+#! commitment checking.
+#!
+#! input: [words, write_ptr, ...]
+#! output: [A, B, C, write_ptr', ...]
+export.move_from_tape_to_memory.0
+  
+  # check if there is an odd number of words
+  # stack: [is_odd, words, write_ptr, ...]
+  dup exec.is_odd
+
+  # copy the flag `is_odd`, this will later on define if padding is required
+  # stack: [is_odd, words, write_ptr, needs_padding, ...]
+  dup movdn.3
+
+  # Make sure `words` is event. This is a requirement for the
+  # `move_from_tape_to_memory_even_words`.
+  # - even words will be `words-0` so result stays even
+  # - odd words will be `words-1` so result becomes event
+  #
+  # stack: [words_even, write_ptr, needs_padding, ...]
+  sub
+
+  # Prepare the capacity word. For rescue prime optimized the first element is
+  # set to `1` when padding is used and `0` otherwse, this is determined by the
+  # `needs_padding` flag, so just copy it.
+  dup.1 push.0.0.0
+
+  # set initial state
+  # stack: [A, B, C, write_ptr, words_even, needs_padding, ...]
+  padw padw
```

Comment:
Is this stack order correct? I don't think `write_ptr` and `words_even` got swapped